packages/peregrine/lib/context/eventing.js

```diff
@@ -0,0 +1,57 @@
+import React, {
+    createContext,
+    useCallback,
+    useContext,
+    useMemo,
+    useRef
+} from 'react';
+import { Observable } from 'zen-observable-ts';
+
+const EVENT_NAME = 'beacon';
+const EventingContext = createContext();
+
+const EventingContextProvider = props => {
+    const { children } = props;
+    const cacheRef = useRef([]);
+
+    const state = useMemo(
+        () =>
+            new Observable(observer => {
+                for (const element of cacheRef.current) {
+                    observer.next(element);
+                }
+
+                document.addEventListener(EVENT_NAME, event => {
+                    observer.next(event.detail);
+                });
+            }),
+        []
+    );
+
+    const dispatch = useCallback(detail => {
+        const event = new CustomEvent(EVENT_NAME, { detail });
```

Comment:
This `dispatch` function supports both but when we `dispatch` from the framework are we just indicating that an event happened (ex: "signed in!"), or are we populating a bunch of contextual data along with it?

Reply:
I would think of these events as similar to Redux actions.

```es6
dispatch({
  type: "PAGE_VIEW",
  payload: {
    categoryId: "MTM=",
    pageType: "CATEGORY",
    pathname: "/default/venia-dresses.html",
    query: "?page=1"
  }
})
```